### PR TITLE
Classification metrics

### PR DESCRIPTION
**Pull request by maximzubkov**

### CODE REVIEW

File: model/code2class.py [hunk 1]
```diff
@@ -24,28 +25,43 @@ def __init__(
...
 
     def training_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
         # [batch size; num_classes]
         logits = self(batch.context, batch.contexts_per_label)
         loss = F.cross_entropy(logits, batch.labels.squeeze(0))
         log = {"train/loss": loss}
-        return {"loss": loss, "log": log}
+        with torch.no_grad():
+            conf_matrix = confusion_matrix(logits.argmax(-1), batch.labels.squeeze(0))
+        log["train/accuracy"] = conf_matrix.trace() / conf_matrix.sum()
```
Comment (SpirinEgor) [inline model/code2class.py:52]: Do this under `no_grad` too.

File: model/code2class.py [hunk 2]
```diff
@@ -24,28 +25,43 @@ def __init__(
...
         return self.classifier(self.encoder(samples), paths_for_label)
 
     def _general_epoch_end(self, outputs: List[Dict], loss_key: str, group: str) -> Dict:
         logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
-        progress_bar = {k: v for k, v in logs.items() if k in [f"{group}/loss"]}
+        accumulated_conf_matrix = torch.zeros(self.num_classes, self.num_classes, requires_grad=False)
+        for out in outputs:
+            _conf_matrix = out[f"{group}/confusion_matrix"]
+            max_class_index, _ = _conf_matrix.shape
+            accumulated_conf_matrix[:max_class_index, :max_class_index] += _conf_matrix
```
Comment (SpirinEgor) [inline model/code2class.py:40]: If labels and predictions don't contain zero class, will the confusion matrix ignore it, or will there be row and column of zeros?

File: model/code2class.py [hunk 3]
```diff
@@ -24,28 +25,43 @@ def __init__(
...
+        logs[f"{group}/accuracy"] = accumulated_conf_matrix.trace() / accumulated_conf_matrix.sum()
```
Comment (SpirinEgor) [inline model/code2class.py:41]: Do it under `no_grad()`

File: model/code2class.py [hunk 4]
```diff
@@ -24,28 +25,44 @@ def __init__(
...
 
     def forward(self, samples: Dict[str, torch.Tensor], paths_for_label: List[int]) -> torch.Tensor:
         return self.classifier(self.encoder(samples), paths_for_label)
 
     def _general_epoch_end(self, outputs: List[Dict], loss_key: str, group: str) -> Dict:
-        logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
-        progress_bar = {k: v for k, v in logs.items() if k in [f"{group}/loss"]}
+        with torch.no_grad():
+            logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
+            accumulated_conf_matrix = torch.zeros(self.num_classes, self.num_classes, requires_grad=False)
```
Comment (SpirinEgor) [inline model/code2class.py:37]: You should pass the device. Use `self.device` for this.

File: model/code2class.py [hunk 5]
```diff
@@ -24,28 +25,44 @@ def __init__(
...
+            for out in outputs:
+                _conf_matrix = out[f"{group}/confusion_matrix"]
+                max_class_index, _ = _conf_matrix.shape
+                accumulated_conf_matrix[:max_class_index, :max_class_index] += _conf_matrix
+                logs[f"{group}/accuracy"] = accumulated_conf_matrix.trace() / accumulated_conf_matrix.sum()
```
Comment (SpirinEgor) [inline model/code2class.py:42]: Calculating accuracy should be done after accumulating all matrices.

File: model/code2class.py [hunk 6]
```diff
@@ -24,28 +25,44 @@ def __init__(
...
         logits = self(batch.context, batch.contexts_per_label)
         loss = F.cross_entropy(logits, batch.labels.squeeze(0))
         log = {"train/loss": loss}
-        return {"loss": loss, "log": log}
+        with torch.no_grad():
+            conf_matrix = confusion_matrix(logits.argmax(-1), batch.labels.squeeze(0))
+            log["train/accuracy"] = conf_matrix.trace() / conf_matrix.sum()
+        progress_bar = {"train/accuracy": log["train/accuracy"]}
+
+        return {"loss": loss, "log": log, "progress_bar": progress_bar, "train/confusion_matrix": conf_matrix}
```
Comment (SpirinEgor) [inline model/code2class.py:56]: I think we can store confusion matrix without specifying group, just like statistics for code2seq model.

File: model/code2class.py [hunk 7]
```diff
@@ -34,12 +34,14 @@ def forward(self, samples: Dict[str, torch.Tensor], paths_for_label: List[int])
     def _general_epoch_end(self, outputs: List[Dict], loss_key: str, group: str) -> Dict:
         with torch.no_grad():
             logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
-            accumulated_conf_matrix = torch.zeros(self.num_classes, self.num_classes, requires_grad=False)
+            accumulated_conf_matrix = torch.zeros(
+                self.num_classes, self.num_classes, requires_grad=False, device=self.device
+            )
             for out in outputs:
-                _conf_matrix = out[f"{group}/confusion_matrix"]
+                _conf_matrix = out["statistic"]
```
Comment (SpirinEgor) [inline model/code2class.py:41]: I meant `confusion_matrix`, just without `train`...
